### DISCUSSION TITLE
What advantages keptn would offer with Argo stack?

### DISCUSSION THREAD

**Discussion by Adiqq** [keptn/keptn#7292] [Q&A]
We're currently using Argo Workflows and Argo CD on our clusters and these components will have to stay. Lately also Argo Events was added.  We're using k3d for local development clusters and we use Rancher based clusters with nginx ingress for staging and production. We use ArgoCD with App of Apps pattern to deploy everything on clusters. I'm looking into keptn concepts and would like to ask about your experience with Argo integrations. Why would you use keptn with Argo stack and how exactly are they complementing each other? 

- How concepts like shipyard, multi-stage delivery, quality gates, automated operations map to Argo Workflows or Argo CD? Would it make sense to use these concepts together with what Argo has to offer or it doesn't make sense, because functionality is overlapping? Would it make sense to integrate e.g. quality gate as Argo workflow step? What about shipyard, would that make sense as Argo workflow step? What potential advantages would be here?
- I saw Istio integration mentioned, what advantages does it offer with keptn? What would we miss with just nginx ingress?

**Comment by agardnerIT** [comment]
@bradmccoydev You've done lots with Argo. Any thoughts?

**Comment by bradmccoydev** [comment]
Hi @Adiqq to answer the last question first:
 - The advantage of using Istio with Keptn is for different deployment strategies such as canary and blue-green, you could still achieve this with Nginx but would have to use something like Argo rollouts to help.
 
 Using Keptn with Argo makes a lot of sense, that is what i use. its a longer question to answer as there are lots of moving parts but perhaps we could show you a demo sometime

**Reply by agardnerIT** [reply]
Thanks Brad. @Adiqq if you choose to use the `helm-service` (which comes "out-of-the-box" with Keptn) AND you choose NOT to use a `user_managed` deployment (whereby the helm-service will just apply your chart as-is). Then there's a reliance on Istio being available on the cluster and thus you can do a blue / green deployment like this:

```
...
tasks:
  - name: deployment
    properties:
      deploymentstrategy: blue_green_service
```

Of course, keptn is completely unopinionated so you're not forced to use the `helm-service` or indeed helm at all to do your deployments.
